JPA 로 전환 - Point

soosue opened the pull request:
closes #12 
### 포인트 적립 내역, 포인트 사용 내역에 대해 JPA 로 전환하기

- PointAccumulate, PointUse 테이블에 대해 엔티티를 구현했습니다.
- 저장하는 로직과 조회하는 로직에 대하여 JPA 를 이용하도록 변경했습니다.
- 조회 화면에 대해서 javascript 로직을 공통화 하였습니다.
- url 을 컨벤션에 맞게 변경하였습니다.
- lombok getter 를 추가하였습니다.
- 테이블의 이름과 칼럼은 그대로 유지하는 방향으로 진행하였습니다. (우선적으로 JPA 로 빠른 전환이 필요하다고 생각했습니다.)

안녕하세요 성현님! 오랜만에 PR 요청드리네요. 피드백 부탁드립니다!
추가적으로 4회차 챌린지를 진행하려면, 도메인 코드가 존재해야할 것 같은데, 현재 상태로는 테스트코드 작성이 쉽지 않을 것 같다는 생각이 드는데요.
정적분석도구에 대한 챌린지를 먼저 진행하고 테스트코드 관련 챌린지는 리팩토링하면서 도메인 로직을 정리한 이후에 진행해야겠다고 생각하고 있습니다. 혹시 좋은 의견이 있으시다면 주시면 감사하겠습니다 :)

soosue committed:
chore: GuestHouseDao 미사용 코드 제거

soosue committed:
fix: 비밀번호 틀려서 member 가 null 일 때, NPE 발생 오류 수정

soosue committed:
refactor: PointAccumulate 엔티티 구현 및 insert JPA 로 전환

soosue committed:
refactor: 체험 예약시 PointAccumulate JPA 로 전환 및 PointAccumulate Type 추가

soosue committed:
chore: 미사용 코드 제거

soosue committed:
refactor: Member 별 포인트 적립 개수 조회 JPA 화

soosue committed:
fix: GenerationType.SEQUENCE 로 변경

soosue committed:
fix: PointAccumulate 테이블명 지정해주기

soosue committed:
refactor: url 변경 및 코드 위치 변경

soosue committed:
refactor: 포인트 관리 페이지로 이동하는 url 변경

soosue committed:
refactor: 포인트 관리 페이지로 이동하는 컨트롤러에선 로직이 불필요하여 제거

soosue committed:
refactor: 기존 pointManage() 메서드가 포인트 적립내역, 포인트 사용내역 모두 조회해서 보여주므로 2개의 메서드로 분리

soosue committed:
refactor: 기존 포인트 적립 내역 조회하는 화면 가독성 올리기

- 포인트적립내역과 포인트사용내역 조회하는 javascript 함수 분리
- url 변경

soosue committed:
refactor: 포인트 적립 조회 리팩토링

- 비동기 조회를 이용한 화면 렌더링
- 포인트 적립 내역만 조회하도록 변경
- PointAccumulate JPA 화
- 불필요한 로직 제거

soosue committed:
chore: 미사용 코드 제거 PointAccumulateDto

soosue committed:
refactor: javascript 로 paging 하여 화면 렌더링 하는 함수 공통화

soosue committed:
refactor: PointUse 를 PointUseDto 로 변경

soosue committed:
refactor: PointUse 엔티티 구현 및 저장로직 JPA 화

soosue committed:
refactor: 포인트 사용 내역 조회 JPA 화

soosue committed:
refactor: Pagination click event 추가하는 함수 공통화

soosue committed:
chore: 미사용 코드 제거

soosue committed:
refactor: 현재 보유포인트 조회하는 로직 추가

soosue committed:
chore: 미사용 코드 제거

soosue committed:
chore: url 경로 변경 .page 를 붙여 페이지 이동을 명시

soosue committed:
chore: 미사용 코드 제거

soosue committed:
chore: PointUse, PointUseRepository 패키지 이동

soosue committed:
chore: 의미가 명확한 메소드명으로 변경

soosue committed:
refactor: lombok getter 추가

boorownie commented on src/main/java/com/java/guesthouse/point/domain/PointAccumulate.java:
```diff
@@ -0,0 +1,56 @@
+import lombok.Getter;
+
+@Entity
+@Table(name = "pointaccumulate")
```
기존 테이블에 새로운 클래스를 맵핑하신거군요!

soosue replied:
네 맞습니다! 테이블이나 칼럼명 개선은 JPA 로 모든 전환이 끝난 이후에 필요하다면 할 생각입니다!

boorownie commented on src/main/java/com/java/guesthouse/point/domain/PointAccumulate.java:
```diff
@@ -0,0 +1,56 @@
+    @Enumerated(EnumType.STRING)
+    private PointType pointType;
+
+    @Column(name = "accudate")
```
테스트와 프로덕션의 환경이 다를 경우 컬럼의 속성으로 인해 의도하지 않은 장애가 발생할 수  있는데요
컬럼의 추가 속성(length 등)을 함께 기재하면 좋을 것 같아요

soosue replied:
환경 다르다는 얘기가 DB 종류가 다른걸 얘기하시는거 맞을까요?

boorownie replied:
네네 맞습니다. DB가 다른경우를 의미했어요

boorownie commented on src/main/java/com/java/guesthouse/point/domain/PointUse.java:
```diff
@@ -0,0 +1,46 @@
+@Getter
+public class PointUse {
+    @Id
+    @GeneratedValue(strategy = GenerationType.SEQUENCE)
```
SEQUENCE를 선택한 이유가 있을까요?

soosue replied:
처음 썼을 때는 오라클을 사용하고 있어서 선택했었습니다.
이유를 아는 것은 아니지만 AUTO 를 쓰면 안되고 명확하게 원하는 키 생성 전략을 선택해서 사용해야 한다고 들었는데요.

그 이유는 찾아봤는데 잘 모르겠습니다.
AUTO 를 사용하면 MYSQL 에서 커넥션을 본 트랜잭션에 1개, ID 채번하는데 1개 총 2개를 소비한다는 글을 봤는데요.
그렇기 때문에 AUTO 를 사용하지 말고 IDENTITY 로 명시해주어야 한다는걸까요? 그럼 오라클 같은 경우엔 어차피 SEQUENCE 를 이용하니, AUTO 로 해도 될 것 같다는 생각이 듭니다. 염치 없지만.. 아시는 게 있으시다면 도와주세요 ㅠㅠ

boorownie replied:
각 설정마다 조금씩 다르니 상황에 맞게 활용하기 위해서 알아보셨으면 해서 남긴 코멘트 입니다.
사실 저는 IDENTITY를 사용하는 편이라서 여쭤본거도 있긴 해요ㅎㅎ

boorownie commented on src/main/java/com/java/guesthouse/point/ui/PointController2.java:
```diff
@@ -0,0 +1,48 @@
+import com.java.guesthouse.point.service.dto.PointUsesResponse;
+
+@Controller
+public class PointController2 {
```
PointController와 무슨 관계인가요?

soosue replied:
PointController 가 old 버전이고
PointController2가 new 버전입니다.

old 버전이 패키지가 이상한 위치에 있고 해서, 따로 구분해서 만들어놓고 필요한 로직들은 이동시키려고 만들었는데, 같은 이름이라 빈 생성시에 충돌이 나서 2를 붙였었습니다... 기존 PointController 를 OldPointController 로 변경하고 2를 PointController 로 가져가보겠습니다.!

boorownie commented on src/main/java/com/java/guesthouse/point/ui/PointController2.java:
```diff
@@ -0,0 +1,48 @@
+        return ResponseEntity.ok(pointService.getPointByMemberId(memberId));
+    }
+
+    @GetMapping("/v1/pointaccumulates/me")
```
pointaccumulates이 카멜케이스 혹은 스네이크케이스 중 하나로 표현하는건 어때요?

soosue replied:
네 저도, 보다보니깐 이게 맞나 싶더라고요. 그래서 - 로 구분을 해보려고 했는데 피드백을 주셨네요!

근데 궁금한게 있습니다. 제안 주신 것 중에서 pointAccumulates 와 point_accumulates 를 주셨는데요.
두 개를 제안 주신 이유가 있을까요?

point-accumulates 는 빠진 이유가 궁금합니다. 찾아보니 [google 검색 가이드](https://developers.google.com/search/docs/crawling-indexing/url-structure?hl=ko) 를 보면 - 을 쓰는걸 장려한다고 하더라고요. 의견 기다리겠습니다!

boorownie replied:
제가 제안드린 두 표현 중 꼭 하나를 선택할 필요는 없었습니다!
사실 하이픈을 이용한 표현 이름이 생각나질 않아서 두가지만 제안드렸어요 ㅎㅎ;;
찾아주신 자료를 보니 하이픈이 더 적절할 수 있을 것 같아요!

boorownie commented on src/main/java/com/java/guesthouse/point/domain/PointUse.java:
```diff
@@ -0,0 +1,46 @@
+
+@Entity
+@Table(name = "pointuse")
+@Getter
```
롬복 추가! 👍

boorownie commented on src/main/java/com/java/guesthouse/point/domain/PointUse.java:
```diff
@@ -0,0 +1,46 @@
+import lombok.Getter;
+
+@Entity
+@Table(name = "pointuse")
```
테이블명 혹은 컬럼 이름을 변경하는 작업은 큰 작업이죠ㅎㅎ

soosue replied:
처음에는 깔끔하게 프로젝트를 완성하고 싶다는 마음에 테이블 명이랑 칼럼명을 변경을 다 하려고 했었는데요...
성현님 말대로 실제 운영상황이라고 가정하고 프로젝트를 진행하면, JPA 로 전환하면서, 이름을 변경하는 일이 쉽지 않더군요.
그래서 얘기해주셨던것처럼 작업을 단순하게 가져가기로 했습니다! 먼저 JPA로 전환 작업을 먼저하고 필요하다면 추후에 이름들을 변경하도록 하려구요 :)

boorownie reviewed (approved):
고생 많으셨습니다!
말씀주신대로 현재 상태에서는 테스트 코드 작성이 쉽지 않을 수 있어요
개인적으로 추천드리는 방법은 전체 기능 테스트 (E2E 테스트 혹은 API 테스트)를 먼저 작성하고
이 테스트로 기능을 보호한 다음 리팩터링을 하면서 테스트를 작성해나가는 방법입니다.

전체 기능 테스트로 보호받는 프로덕션 코드는 리팩터링 중 사이드 이펙트가 발생하더라도 캐치할 확률이 높습니다. 또, 리팩터링 시 테스트 코드로 보호할 코드를 정한 후 리팩터링 하는 방향으로 한다면 조금 느리더라도 안정적인 리팩터링과 테스트 코드 작성을 기대할 수 있어요.

추가적으로 정적분석도구 적용 순서는 테스트에 대한 챌린지의 진행 상황과 무관하게 진행해도 된다고 생각합니다.! 고생 많으셨어요~

soosue commented:
> 개인적으로 추천드리는 방법은 전체 기능 테스트 (E2E 테스트 혹은 API 테스트)를 먼저 작성하고
이 테스트로 기능을 보호한 다음 리팩터링을 하면서 테스트를 작성해나가는 방법입니다.

>전체 기능 테스트로 보호받는 프로덕션 코드는 리팩터링 중 사이드 이펙트가 발생하더라도 캐치할 확률이 높습니다. 또, 리팩터링 시 테스트 코드로 보호할 코드를 정한 후 리팩터링 하는 방향으로 한다면 조금 느리더라도 안정적인 리팩터링과 테스트 코드 작성을 기대할 수 있어요.

피드백 주셔서 고민을 해봤는데요. 현재 프로젝트가 JSP 로 되어있어서 E2E 테스트를 어떻게 해야하나 찾아보니 Selenium 같은 걸로 코드로 태그를 클릭하고 하는 식의 테스트가 가능한 것 같더라고요.

하지만 이 부분은 단기간에 적용하기가 쉽지 않을 것 같고, 현재 프로젝트 리팩토링에 있어 Selenium 을 통해서 테스트 하는 방식을 익히는 것이 의미가 있는 것일까 생각했습니다.

그래서 생각한 대안으로 mock 을 이용하여 service 에서 행위를 검증하는 방법이 있는 것 같아서 이를 이용해보려고 하는데요. 어떻게 생각하시는지 궁금합니다. 제가 생각했던 행위 검증은 이 글에 나와있고, 글처럼 mock 을 이용한 테스트 부분이면 활용할 수 있을 것 같았습니다.
https://martinfowler.com/articles/mocksArentStubs.html#TheDifferenceBetweenMocksAndStubs 에서 `Using mocks this test would look quite different.` 를 검색하시면 바로 보실 수 있습니다.

soosue committed:
refactor: sonarcloud: resolves bugs, code smells

- memberCode 코드 중복 제거
- 시각장애인들을 위한 table 설명달기
- deprecated 된 속성 제거(align, width, height)

soosue committed:
chore: 미사용 코드 제거

soosue committed:
chore: 중복코드 메서드 분리
